src/main/java/gregtech/common/metatileentities/electric/MetaTileEntityFloodGate.java

```diff
@@ -0,0 +1,306 @@
+    }
+
+    @Override
+    protected IItemHandlerModifiable createExportItemHandler() {
+        return new ItemStackHandler(1);
+    }
+
+    @Override
+    public <T> T getCapability(Capability<T> capability, EnumFacing side) {
+        return (side == null || side.getAxis() != EnumFacing.Axis.Y) ? super.getCapability(capability, side) : null;
+    }
+
+    private boolean isFluid(BlockPos pos) {
+        Block block = getWorld().getBlockState(pos).getBlock();
+        return block instanceof IFluidBlock || block instanceof BlockLiquid;
+    }
+
+    private boolean canPutAt(BlockPos pos) {
+        if(getWorld().isAirBlock(pos)) {
+            return true;
+        }
+
+        if(isFluid(pos)) {
+            IFluidHandler fluidHandler = FluidUtil.getFluidHandler(getWorld(), pos, null);
+            FluidStack drainStack = null;
+
+            if (fluidHandler != null)
+                drainStack = fluidHandler.drain(Integer.MAX_VALUE, false);
+            if (drainStack != null)
+                return drainStack.amount < Fluid.BUCKET_VOLUME;
+
+            return true;
+        }
+
+        return getWorld().getBlockState(pos).getBlock().isReplaceable(getWorld(), pos);
+    }
+
+    private void moveHeadToBottom() {
+        BlockPos selfPos = getPos();
+
+        blocksToCheck.setCenter(selfPos);
+        airBlocks.setCenter(selfPos);
+        checkedBlocks.setCenter(selfPos);
+
+        int temp = 0;
+        while(canPutAt(selfPos.down(temp + 1)))
+            temp ++;
+        if(isFluid(selfPos.down(temp + 1)) && (!isCompleted))
+            temp ++;
+
+        if(headY != temp) {
+            headY = temp;
+            writeCustomData(200, b -> b.writeVarInt(headY));
+            markDirty();
+        }
+    }
+
+    @Override
+    protected ModularUI createUI(EntityPlayer entityPlayer) {
+        TankWidget tankWidget = new TankWidget(importFluids.getTankAt(0), 69, 52, 18, 18)
+            .setHideTooltip(true)
+            .setAlwaysShowFull(true);
+
+        ModularUI.Builder builder = ModularUI.defaultBuilder()
+            .image(7, 16, 81, 55, GuiTextures.DISPLAY)
+            .widget(tankWidget)
+            .label(11, 20, "gregtech.gui.fluid_amount", 0xFFFFFF)
+            .dynamicLabel(11, 30, tankWidget::getFormattedFluidAmount, 0xFFFFFF)
+            .dynamicLabel(11, 40, tankWidget::getFluidLocalizedName, 0xFFFFFF)
+            .label(6, 6, getMetaFullName())
+            .widget(new FluidContainerSlotWidget(importItems, 0, 90, 17, false)
+                .setBackgroundTexture(GuiTextures.SLOT, GuiTextures.IN_SLOT_OVERLAY))
+            .widget(new ImageWidget(91, 36, 14, 15, GuiTextures.TANK_ICON))
+            .widget(new SlotWidget(exportItems, 0, 90, 54, true, false)
+                .setBackgroundTexture(GuiTextures.SLOT, GuiTextures.OUT_SLOT_OVERLAY))
+            .bindPlayerInventory(entityPlayer.inventory);
+        return builder.build(getHolder(), entityPlayer);
+    }
+
+    private void updateQueueState() {
```

Comment:
This function entirely won't work with unlimited radius.
Needs rewriting, most probably via grouping queue by chunks to make filling easier and checking chunk loaded/unloaded faster.